main/src/ca/uwaterloo/flix/language/phase/Codegen.scala

```diff
@@ -1093,8 +1093,18 @@ object Codegen {
           case BinaryOperator.NotEqual => (IF_ICMPEQ, FCMPG, DCMPG, IFEQ)
         }
         e1.tpe match {
-          case Type.Unit | Type.Str if o == BinaryOperator.Equal || o == BinaryOperator.NotEqual =>
-            // Unit and String can be compared using Object's `equal` method
+          case Type.Unit if o == BinaryOperator.Equal || o == BinaryOperator.NotEqual =>
+            // Unit can only be equal to unit, so the objects are poped from the top of the stack
+            visitor.visitInsn(POP)
+            visitor.visitInsn(POP)
+            (e2.tpe, o) match {
+              case (Type.Unit, BinaryOperator.NotEqual) => visitor.visitJumpInsn(GOTO, condElse)
+              case (Type.Unit, BinaryOperator.Equal) =>
+              case (_, BinaryOperator.Equal) => visitor.visitJumpInsn(GOTO, condElse)
+              case (_, BinaryOperator.NotEqual) =>
+            }
```

Comment:
Yes, and it would be good to have a comment in each branch. E.g.:

* A unit value is always equal itself, so no need to branch.
* A  unit value is never unequal to itself, so always branch.

or something